### ISSUE TITLE
[Coroutines] Invalided symmetric transfer

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#52896]
See https://godbolt.org/z/E55h9Pv9f , the symmetric transfer is invalidated so that the example would be stack-overflowing, which should be avoided by the symmetric transfer optimization.

BTW, it is OK in 12: https://godbolt.org/z/8Y8YbM8eK. This is introduced in https://github.com/llvm/llvm-project/commit/822b92aae439c4ba2946980c8a27bd2c8a62d90c#diff-8cf21bda84c593733aa099f89fe7d197fd83203c6bc4f6fbdcd94f8bb4256d23L1138-L1147 , which deletes some optimization passes which symmetric transfer depends on.

**Comment by iains** [comment]
since symmetric transfer (i.e. an indirect tail call) can not be guaranteed for all targets (several common ABIs do not support an indirect tail call).

I think that we need a design agreed between in the compiler 'vendors' (probably with a builtin function for the continuation).  There is a nominal coroutines "ABI" group and perhaps that needs to be convened to discuss this.

A skeleton idea;
 - implement the continuation calls as a builtin
 - if the target unconditionally supports an indirect tail call then the builtin can be lowered to that.
 - if the target needs registers reloading for cross-DSO jumps, then I would suggest that we use reserved space below the existing frame pointers is that it can be found at a small offset from the frame pointer.  It would seem sad to make a trampoline when we already have somewhere to save the data.

```
  * so here *
--> frame pointe
  (*resume)(*frame)
 (*destroy) (*frame)
 promise...
 ...
```

**Comment by iains** [comment]
this is GCC BZ: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=94794

@GorNishanov @zygoloid  - thoughts?
(whatever is done, it would be better to coordinate the design)/

**Comment by ChuanqiXu9** [comment]
Hi @iains , thanks for looking into this! I have made a patch for this issue: https://reviews.llvm.org/D116327. I posted this as an issue since I think this could be beneficial to other people who are not clang/llvm developers.
I fully agree that it is a problem really. I've seen 2 issues related to this (https://github.com/llvm/llvm-project/issues/42853 and https://github.com/llvm/llvm-project/issues/47971).
Finally, I think it might be better to discuss this in cfe-dev/llvm-dev or another issue page for compatible coroutines "ABI" group. Since I guess this page would be closed if every thing  goes well.

(BTW, does the coroutines "ABI" group work actively recently? Could I join it? How could I?)

**Comment by ChuanqiXu9** [comment]
This should be addressed at: https://github.com/llvm/llvm-project/commit/403772ff1ce5618c8d02316531386b415312274a

**Comment by ChuanqiXu9** [comment]
@tstellar since this is a regression bug, I feel it is necessary to backport to 13.x. I've seen your proposal to use New Automated Release Workflow (using Issues and Pull Requests) in: https://lists.llvm.org/pipermail/llvm-dev/2021-December/154389.html. But it says it would work since 14.x. So if we want to backport this to 13.x, what should I do here?

**Comment by tstellar** [comment]
@ChuanqiXu9 I can take care of it.

**Comment by tstellar** [comment]
Merged: b9a243d1cac2
